### PR TITLE
test: prove missing InstantSend quorum is reachable from peers [v2]

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented

A peer can send a structurally valid ISDLOCK with the genesis block as its cycle hash and an arbitrary parseable BLS signature. The receive path accepts the known cycle-boundary hash, but a mature mainnet node selects signing height 287, before DIP0024 activation, so no quorum is available. The current early return then discards the entire dequeued batch, including unrelated locks that could otherwise be processed.

Current regression tests asserts current implementation and testing environment is very much spefic on the current implementation rather than really testing behaviour.

## What was done?

Regression test is removed and added a new functional test's scenario with a genuine lock that shares a batch with the poisoned one still gets applied by reverting commit 71cf6252599d1e716e0f5e68628eb9cddd935069 and partial revert for 2e7f4a9a37894b28cdfb288a7087e9cbd4964cab

## How Has This Been Tested?
It runs 5 attemps just to be sure that it doesn't get to the different batch and slip out from validation by mistake. Asserts the target reports instantlock true for the genuine transaction, and that the peer stays connected.

Tested by reverting changes from #7662 - functional tests fails as expected:

        2026-09-12T10:55:16.127000Z TestFramework (INFO): A missing-quorum ISDLOCK must not drop the genuine locks batched with it
    2026-09-12T10:55:17.541000Z TestFramework (INFO): Expecting InstantLock for ['dd3f8e9b6b0d2db0eefae805169a5f01d37acf615739f7cdaf35fa8f117097d5']
    2026-09-12T10:55:37.656000Z TestFramework.utils (ERROR): wait_until() failed. Predicate: ''''
                self.wait_until(lambda txid=txid: target.getrawtransaction(txid, True)["instantlock"], timeout=20)
    '''
    2026-09-12T10:55:37.656000Z TestFramework (ERROR): Assertion failed
    Traceback (most recent call last):
      File "/DASH/test/functional/test_framework/test_framework.py", line 164, in main
        self.run_test()
        ~~~~~~~~~~~~~^^
      File "/DASH/test/functional/p2p_instantsend.py", line 63, in run_test
        self.test_genesis_cycle_islock_isolated()
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
      File "/DASH/test/functional/p2p_instantsend.py", line 205, in test_genesis_cycle_islock_isolated
        self.wait_until(lambda txid=txid: target.getrawtransaction(txid, True)["instantlock"], timeout=20)
        ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/DASH/test/functional/test_framework/test_framework.py", line 908, in wait_until
        return wait_until_helper(test_function, timeout=timeout, lock=lock, timeout_factor=self.options.timeout_factor, sleep=sleep, do_assert=do_assert)
      File "/DASH/test/functional/test_framework/util.py", line 298, in wait_until_helper
        raise AssertionError("Predicate {} not true after {} seconds".format(predicate_source, timeout))
    AssertionError: Predicate ''''
                self.wait_until(lambda txid=txid: target.getrawtransaction(txid, True)["instantlock"], timeout=20)
    ''' not true after 20 seconds
    2026-09-12T10:55:37.708000Z TestFramework (INFO): Stopping nodes



## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by thepastaclaw** [comment]
🕓 Review not started yet because this PR is a draft.

- [ ] **Request normal review** — click when the PR is ready for review.
- [ ] **Request priority review** — click to move this review to the front of the queue.

_Commit ec683243. Normal review starts when eligible; priority review starts as soon as a slot is available._

File: test/functional/p2p_instantsend.py [hunk 1]
```diff
@@ -171,6 +175,38 @@ def test_isdlock_relayed_to_recsigs_observer(self):
...
+            genuine = msg_isdlock()
+            genuine.deserialize(BytesIO(bytes.fromhex(controller.getislocks([txid])[0]["hex"])))
+            # Reusing a real signature keeps sig.IsValid() true so the crafted lock reaches quorum selection.
+            poisoned = msg_isdlock(1, [COutPoint(random.getrandbits(256), 0)], random.getrandbits(256),
+                                   genesis_hash, genuine.sig)
+            target.sendrawtransaction(controller.getrawtransaction(txid))
+            peer.send_message(poisoned)
+            peer.send_message(genuine)
+            peer.sync_with_ping()
+            self.wait_until(lambda txid=txid: target.getrawtransaction(txid, True)["instantlock"], timeout=20)
```
Comment (thepastaclaw) [inline test/functional/p2p_instantsend.py:205]: **🟡 Suggestion: Establish the shared-batch condition before relying on this regression test**

The two sends do not establish that the poisoned and genuine locks share a verification batch. `send_message()` schedules separate asynchronous socket writes, and `NetInstantSend::WorkThreadMain()` independently calls `FetchPendingLocks()`, which removes the currently queued locks under `cs_pendingLocks`. The worker can fetch the poisoned lock before the genuine lock is enqueued; with the old `return nullptr`, only the poisoned lock's batch is discarded, and the genuine lock succeeds in the next batch. `sync_with_ping()` synchronizes network message handling, not the InstantSend worker. Five attempts and the reported failure with the fix reverted provide useful empirical evidence, but do not exclude a false pass under a different schedule. Since this PR also removes the deterministic `missing_quorum_does_not_drop_batch` unit test, establish or observe that both locks are processed in one batch, or retain deterministic batch-level coverage alongside this functional scenario.

<sub>source: `gpt-6-astra` (phase2-reviewer: general)</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Phase 2 only (queue backlog)

Verified the supplied Phase-2 finding against head 7c5a3746cc4107ceabb1d8242eabb5960c81e62f. The replacement functional scenario exercises peer delivery, but its shared-batch precondition remains scheduling-dependent while this PR removes deterministic batch-level coverage. The author's reported failure with the fix reverted demonstrates detection in that run, not a guarantee across schedules; this is a test-coverage suggestion, with no blocking findings.

🟡 1 suggestion(s)

<details>
<summary>Review provenance</summary>

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

- Triage: `normal` by `gpt-6-astra` (effort low) — This is a focused regression-test change spanning InstantSend functional testing, removal of a unit test, and a header adjustment, but the diff itself does not alter production consensus, cryptography, peer deserialization, or other critical behavior.
- Phase 1 reviewers: **not run (skipped for throughput: 12 PRs queued, above the 10 limit)**
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `test/functional/p2p_instantsend.py`:
- [SUGGESTION] test/functional/p2p_instantsend.py:202-205: Establish the shared-batch condition before relying on this regression test
  The two sends do not establish that the poisoned and genuine locks share a verification batch. `send_message()` schedules separate asynchronous socket writes, and `NetInstantSend::WorkThreadMain()` independently calls `FetchPendingLocks()`, which removes the currently queued locks under `cs_pendingLocks`. The worker can fetch the poisoned lock before the genuine lock is enqueued; with the old `return nullptr`, only the poisoned lock's batch is discarded, and the genuine lock succeeds in the next batch. `sync_with_ping()` synchronizes network message handling, not the InstantSend worker. Five attempts and the reported failure with the fix reverted provide useful empirical evidence, but do not exclude a false pass under a different schedule. Since this PR also removes the deterministic `missing_quorum_does_not_drop_batch` unit test, establish or observe that both locks are processed in one batch, or retain deterministic batch-level coverage alongside this functional scenario.
```
</details>

File: test/functional/p2p_instantsend.py (hunk 1 above)
Comment (PastaPastaPasta) [reply test/functional/p2p_instantsend.py:205]: @knst
this seems like a valid concern to me; maybe even more of a blocker

Comment (thepastaclaw) [reply test/functional/p2p_instantsend.py:205]: **Still applies** (re-reviewed at `7c5a3746`): This remains unresolved at the reviewed head: the two asynchronous sends and subsequent ping do not synchronize with `FetchPendingLocks()`, so the test still cannot prove that both locks entered one batch. Removing the deterministic unit test leaves the regression dependent on scheduling and permits a false pass on the unfixed implementation.

Comment (knst) [reply test/functional/p2p_instantsend.py:205]: @PastaPastaPasta the tests should test functionality, not the very specific implementation. existing regression test is exact case when _very specific_ implementation is tested, but not a scenario.
And it should be reliable / error prune / agnostic for any further refactoring, removing-adding locks, queues, etc. 

if @thepastaclaw have suggestion how exactly improve _my_ version of testing in functional tests to make it even more reliable without having expectation from implementation, he is welcome to do it.
but this regression test for the single very specific `if` condition for very specific implementation of queue is extremely unreliable and fragile regression test, it makes extra burden for support in the future.

----
Please check also "How Has This Been Tested?" chapter - with reverted changes from 7662 the new scenario in functional test failed from the 1st attempt [I have 5 of them to increase probability].

Comment (thepastaclaw) [reply test/functional/p2p_instantsend.py:205]: **Still applies** (re-reviewed at `7c5a3746`): The test flow is unchanged with respect to the reported race: two asynchronous sends are followed by `sync_with_ping()`, which does not synchronize `FetchPendingLocks()`. Removing the deterministic unit test leaves no guarantee that the genuine lock shares a batch with the poisoned lock, so the finding remains applicable.

Comment (PastaPastaPasta) [reply test/functional/p2p_instantsend.py:205]: I don't really have a SUPER strong opinion here @knst; but all 3 review bots disagree with you, and my read also leans towards "what's wrong with the unit test" soooo; maybe @UdjinM6 can break the tie for you :D

Comment (thepastaclaw) [reply test/functional/p2p_instantsend.py:205]: **Still applies** (re-reviewed at `7c5a3746`): The current test still sends the two locks through asynchronous socket writes and only waits for a ping response, which does not synchronize the InstantSend worker or prove shared-batch processing. Removing the deterministic unit test leaves this unresolved, so the regression can still pass against the pre-fix implementation under a different scheduling interleaving.

Comment (knst) [reply test/functional/p2p_instantsend.py]: @thepastaclaw see [ec68324](https://github.com/dashpay/dash/pull/7675/commits/ec683243c767f7f9c04c66ee9ad647a6d0b3ac54)

**Comment by knst** [comment]
btw,
I did a benchmark:

   Enqueue gaps between the three locks, versus the 100ms poll:

```
  ┌─────────┬────────┬───────┬───────┐
  │   gap   │ median │ mean  │  max  │
  ├─────────┼────────┼───────┼───────┤
  │ P1 to L │ 127us  │ 171us │ 586us │
  ├─────────┼────────┼───────┼───────┤
  │ L to P2 │ 43us   │ 63us  │ 297us │
  └─────────┴────────┴───────┴───────┘
```

  Batches per attempt:

```
  ┌──────────────────────┬──────────┐
  │       batching       │ attempts │
  ├──────────────────────┼──────────┤
  │ [P L P] in one batch │ 59       │
  ├──────────────────────┼──────────┤
  │ [P] + [L P]          │ 1        │
  ├──────────────────────┼──────────┤
  │ [P L] + [P]          │ 0        │

```

so, in 2% cases they are picked in non-one batch in practice

File: test/functional/p2p_instantsend.py [hunk 2]
```diff
@@ -171,6 +175,44 @@ def test_isdlock_relayed_to_recsigs_observer(self):
...
+            # The worker polls the pending queue every 100ms but we enqueue these three in
+            # microseconds, so at most one poll can split them and every split still batches the
+            # genuine lock with a quorum-less neighbour. The genuine lock therefore cannot reach a
+            # batch of its own, which is what let it survive before the fix.
+            for lock in (genesis_lock(genuine.sig), genuine, genesis_lock(genuine.sig)):
+                peer.send_message(lock)
+            peer.sync_with_ping()
+            self.wait_until(lambda txid=txid: target.getrawtransaction(txid, True)["instantlock"], timeout=20)
+
+        target.disconnect_p2ps()
```
Comment (thepastaclaw) [inline test/functional/p2p_instantsend.py:213]: **💬 Nitpick: Explicitly check peer connectivity after processing the final iteration**

sync_with_ping() checks connectivity before the RPC wait for asynchronous InstantSend verification. That RPC wait does not inspect the peer, and disconnect_p2ps() also succeeds when the peer has already disconnected. Subsequent iterations provide another connectivity check, but the final iteration has none after verification. Add an explicit post-verification connectivity check before deliberate teardown so the test covers the PR description's claim that the peer stays connected.

<sub>source: `gpt-6-astra` (phase2-reviewer: general)</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Re-review — Final validation — Phase 2 only (queue backlog)

Verified the complete diff at ec683243c767f7f9c04c66ee9ad647a6d0b3ac54 against the queue, verification, and P2P helper implementations. The functional scenario improves integration coverage, but shared-batch coverage remains timing-dependent and the final post-verification connectivity check is missing; the introductory commit also describes a log assertion that is not present. Python syntax parsing and diff whitespace checks passed; runtime tests were not run, and no blocking findings were confirmed.

🟡 1 suggestion(s) | 💬 2 nitpick(s)

### 1 finding(s) not shown inline (the lines are not part of this PR's diff)

**💬 Nitpick: Remove the stale log-assertion explanation from the commit message**
`<commit:7c5a3746cc>:1`

The commit message says that the absence of 'doing verification on old active set' makes an assertion fail. Neither this commit's added scenario nor the final version asserts on that log phrase; the scenario waits for the genuine transaction's instantlock result. Update the message to describe the assertion actually introduced so future readers do not infer coverage that the commit does not provide.

<sub>source: `gpt-6-astra` (phase2-reviewer: general, dash-core-commit-history)</sub>


_1 carried-forward finding(s) already raised on this PR; not re-posting as new inline comments._

<details>
<summary>Review provenance</summary>

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); reviewer 3: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 4: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

- Triage: `normal` by `gpt-6-astra` (effort low) — The change replaces unit coverage with a timing-sensitive, multi-node functional regression scenario requiring careful validation of batching and isolation, but does not change critical production behavior.
- Phase 1 reviewers: **not run (skipped for throughput: 32 PRs queued, above the 10 limit)**
- Fresh final gate: an independent Phase-2 review ran after iterative findings were reconciled
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `test/functional/p2p_instantsend.py`:
- [NITPICK] test/functional/p2p_instantsend.py:210-213: Explicitly check peer connectivity after processing the final iteration
  sync_with_ping() checks connectivity before the RPC wait for asynchronous InstantSend verification. That RPC wait does not inspect the peer, and disconnect_p2ps() also succeeds when the peer has already disconnected. Subsequent iterations provide another connectivity check, but the final iteration has none after verification. Add an explicit post-verification connectivity check before deliberate teardown so the test covers the PR description's claim that the peer stays connected.
- [SUGGESTION] test/functional/p2p_instantsend.py:204-209: Establish the shared-batch condition before relying on this regression test
  (existing thread: https://github.com/dashpay/dash/pull/7675#discussion_r3997802936)
  The additional neighbouring lock handles a single batch split, but the comment's guarantee still does not follow from the implementation. send_message() schedules asynchronous socket writes through call_soon_threadsafe(); submitting those writes quickly does not bound when the node enqueues them. The InstantSend worker independently fetches the pending queue, so scheduling delays can still leave the genuine lock in a separate verification batch. The benchmark and reverted-fix failures support practical reliability, but do not establish that every passing run exercises the batch-level regression. Keeping the functional scenario independent of queue implementation is a reasonable goal; however, this PR also removes the deterministic missing_quorum_does_not_drop_batch coverage. Retain focused batch-level coverage alongside the functional scenario, or observe that the required shared-batch condition occurred before counting the run as regression coverage.

In `<commit:7c5a3746cc>`:
- [NITPICK] <commit:7c5a3746cc>:1: Remove the stale log-assertion explanation from the commit message
  The commit message says that the absence of 'doing verification on old active set' makes an assertion fail. Neither this commit's added scenario nor the final version asserts on that log phrase; the scenario waits for the genuine transaction's instantlock result. Update the message to describe the assertion actually introduced so future readers do not infer coverage that the commit does not provide.
```
</details>

File: test/functional/p2p_instantsend.py (hunk 1 above)
Comment (thepastaclaw) [reply test/functional/p2p_instantsend.py]: **Still applies** (re-reviewed at `ec683243`): The change in ec68324 usefully handles any single split, and your benchmark supports its practical reliability, but asynchronous delivery and message processing do not guarantee the sub-100-ms enqueue interval assumed by the comment. Keeping focused deterministic coverage alongside the functional scenario, or observing that the required shared batch occurred, would close the remaining coverage gap without requiring the functional test to control the queue implementation.

Comment (UdjinM6) [reply test/functional/p2p_instantsend.py]: How about making it deterministic? This would allow us to add even more stable tests for various edge cases in the future. See https://github.com/UdjinM6/dash/commit/7c19184c849c382d6c72d31baeced0bd4631c616 for the feat/fix + https://github.com/UdjinM6/dash/commit/bf8b636c0d63ca9f9a3ba65d130ffa3e4df5af88 for some exta test cases

Comment (thepastaclaw) [reply test/functional/p2p_instantsend.py]: **Still applies** (re-reviewed at `ec683243`): Your ec683243c7 sandwich and timing measurements substantially reduce the likelihood of a false pass, but they do not prevent two worker polls from separating delivery under scheduling delays. You can preserve the behavioral approach by pausing verification until delivery is acknowledged and then resuming it, or retain deterministic batch coverage alongside this scenario.

**Review by thepastaclaw — COMMENTED** [review]
## Re-review — Final validation — Phase 2 only (queue backlog)

The functional scenario is a useful behavioral regression test, but its core shared-batch precondition remains timing-dependent: the three asynchronous P2P writes are not proven to enter one InstantSend verification batch. The final iteration also has no connectivity check after the genuine lock is processed, and the introducing commit contains a stale explanation of a log assertion that the test does not implement.

🟡 1 suggestion(s) | 💬 2 nitpick(s)

### 1 finding(s) not shown inline (the lines are not part of this PR's diff)

**💬 Nitpick: Remove the stale log-assertion explanation from the commit message**
`<commit:7c5a3746cc>:1`

The commit message says the test asserts the presence or absence of the log phrase `doing verification on old active set`, but the committed test contains no such log assertion. Its actual regression oracle is the genuine transaction's `instantlock` result. Update the permanent commit message to describe the implemented oracle rather than the removed test approach.

<sub>source: `gpt-6-astra` (phase2-reviewer: general, dash-core-commit-history)</sub>


_2 carried-forward finding(s) already raised on this PR; not re-posting as new inline comments._

<details>
<summary>Review provenance</summary>

Source: reviewer 1: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 2: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); reviewer 3: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 4: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

- Triage: `normal` by `gpt-6-astra` (effort low) — The diff replaces unit coverage with a peer-driven functional regression scenario requiring reasoning about node isolation, quorum selection, and asynchronous batching, but changes no production behavior or critical surface.
- Phase 1 reviewers: **not run (skipped for throughput: 13 PRs queued, above the 10 limit)**
- Fresh final gate: an independent Phase-2 review ran after iterative findings were reconciled
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `<commit:7c5a3746cc>`:
- [NITPICK] <commit:7c5a3746cc>:1: Remove the stale log-assertion explanation from the commit message
  The commit message says the test asserts the presence or absence of the log phrase `doing verification on old active set`, but the committed test contains no such log assertion. Its actual regression oracle is the genuine transaction's `instantlock` result. Update the permanent commit message to describe the implemented oracle rather than the removed test approach.

In `test/functional/p2p_instantsend.py`:
- [SUGGESTION] test/functional/p2p_instantsend.py:204-209: Establish the shared-batch condition before relying on this regression test
  (existing thread: https://github.com/dashpay/dash/pull/7675#discussion_r3997802936)
  The poisoned–genuine–poisoned sequence reduces the chance of a false pass, but it does not establish that all three locks enter the same verification batch. `send_message()` schedules asynchronous socket writes, while the InstantSend worker independently polls and drains pending locks. If the worker runs after the genuine lock is enqueued but before the trailing poisoned lock arrives, the genuine lock can be processed successfully on its own and the pre-fix implementation can still pass. The five repetitions and observed failure with the fix reverted are useful empirical evidence, but they do not exclude this scheduling interleaving. Add a test-visible synchronization mechanism that proves the locks share one verification batch, or retain deterministic batch-level coverage alongside this functional test.
- [NITPICK] test/functional/p2p_instantsend.py:210-213: Explicitly check peer connectivity after processing the final iteration
  (existing thread: https://github.com/dashpay/dash/pull/7675#discussion_r4020972281)
  `sync_with_ping()` checks connectivity before asynchronous InstantSend verification has necessarily completed. The subsequent wait only observes the transaction's `instantlock` field, and `disconnect_p2ps()` does not distinguish a peer that remained connected from one that disconnected earlier. Earlier iterations receive another connectivity-sensitive operation on the next loop, but the final iteration does not. Add a final ping or explicit connection assertion after the genuine lock is observed to cover the stated requirement that the peer remains connected.
```
</details>
